File: generated_code/cpp/solution_203_2.cpp
#include <iostream>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    /**
     * Algorithm: Dummy Head Pointer
     * Time Complexity: O(N) where N is the number of nodes in the list.
     * Space Complexity: O(1) as we modify the list in-place without extra storage.
     * 
     * Using a dummy head simplifies edge cases where the head itself needs to be removed.
     */
    ListNode* removeElements(ListNode* head, int val) {
        // Create a dummy node that points to the head of the list.
        // This handles cases where the original head contains 'val'.
        ListNode dummy(0, head);
        ListNode* curr = &dummy;

        while (curr->next != nullptr) {
            if (curr->next->val == val) {
                // Node needs to be removed.
                ListNode* temp = curr->next;
                curr->next = curr->next->next;
                // In a competitive programming context, we assume memory is managed 
                // by the environment or ignored, but in production C++, we delete it.
                delete temp; 
            } else {
                // Node is valid, move the pointer forward.
                curr = curr->next;
            }
        }

        return dummy.next;
    }
};

File: generated_code/python/solution_203_2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        # Using a dummy node simplifies edge cases where the head itself needs to be removed.
        # Time Complexity: O(N) where N is the number of nodes, as we visit each node exactly once.
        # Space Complexity: O(1) auxiliary space, as we only modify pointers in place.
        dummy = ListNode(next=head)
        current = dummy
        
        while current.next:
            if current.next.val == val:
                # Bypass the node that matches the target value
                current.next = current.next.next
            else:
                # Advance the pointer only if we didn't remove a node
                current = current.next
                
        return dummy.next